### PR TITLE
[draft] Cleaner (?) state machine

### PR DESCRIPTION
**Pull request by TApplencourt**

### CODE REVIEW

File: template/filter.c.erb [hunk 1]
```diff
@@ -137,106 +108,123 @@ filter_message_iterator_next_processing_reading(
...
+      message_iterator_private_data->head_mi->message_iterator);
+  struct el_mi *tmp = message_iterator_private_data->head_mi;
+  CDL_DELETE(message_iterator_private_data->head_mi, tmp);
+  free(tmp);
+
+  return message_iterator_private_data->head_mi;
+}
+
+static inline bt_message_iterator_class_next_method_status
+filter_message_iterator_transition_from_reading_to_ending(
```
Comment (Kerilk) [inline template/filter.c.erb:129]: Should rename to:
`filter_message_iterator_transition_from_reading_to_finalizing`

File: template/filter.c.erb [hunk 2]
```diff
@@ -19,116 +19,87 @@ static char *get_port_name(uint64_t current) {
...
 
-  return BT_MESSAGE_ITERATOR_CLASS_NEXT_METHOD_STATUS_AGAIN;
+  btx_downstream_move_messages(message_iterator_private_data, messages,
+                               capacity, count);
+  return BT_MESSAGE_ITERATOR_CLASS_NEXT_METHOD_STATUS_OK;
 }
 
 static inline bt_message_iterator_class_next_method_status
-filter_message_iterator_next_processing_sending(
+filter_message_iterator_next_ending(
```
Comment (Kerilk) [inline template/filter.c.erb:40]: should be called `filter_message_iterator_next_finalizing`

File: template/filter.c.erb [hunk 3]
```diff
@@ -137,106 +108,123 @@ filter_message_iterator_next_processing_reading(
...
+  btx_push_messages_stream_end(self_message_iterator,
+                               message_iterator_private_data);
+  message_iterator_private_data->state = BTX_FILTER_STATE_FINALIZING;
+  return filter_message_iterator_next_ending(self_message_iterator, messages,
+                                             capacity, count);
 }
 
 static inline bt_message_iterator_class_next_method_status
-filter_message_iterator_next_processing(
+filter_message_iterator_next_reading(
```
Comment (Kerilk) [inline template/filter.c.erb:145]: Should be called `filter_message_iterator_next_processing`

File: template/filter.c.erb [hunk 4]
```diff
@@ -19,116 +19,87 @@ static char *get_port_name(uint64_t current) {
 /*
  *  Handle Upstream Messages
  */
-
 static inline bt_message_iterator_class_next_method_status
-filter_message_iterator_next_initializing(
+filter_message_iterator_next_sending(
```
Comment (Kerilk) [inline template/filter.c.erb:23]: Should be called `filter_message_iterator_next_processing_sending`

File: template/filter.c.erb [hunk 5]
```diff
@@ -137,106 +108,123 @@ filter_message_iterator_next_processing_reading(
...
-  if (message_iterator_private_data->queue) {
-    message_iterator_private_data->processing_state =
-        BTX_FILTER_PROCESSING_STATE_SENDING;
-    /* optimization */
-    return filter_message_iterator_next_processing_sending(
-        self_message_iterator, messages, capacity, count);
-  } else
-    return BT_MESSAGE_ITERATOR_CLASS_NEXT_METHOD_STATUS_AGAIN;
+static inline bool
+btx_closing_port(btx_message_iterator_t *message_iterator_private_data) {
```
Comment (Kerilk) [inline template/filter.c.erb:118]: Should this one be prefixed by `btx_` as it is not part of the metababel library?

File: template/filter.c.erb [hunk 6]
```diff
@@ -253,13 +241,11 @@ filter_message_iterator_next(bt_self_message_iterator *self_message_iterator,
...
-                                                   messages, capacity, count);
+    return filter_message_iterator_next_reading(self_message_iterator, messages,
+                                                capacity, count);
   case BTX_FILTER_STATE_FINALIZING:
-    return filter_message_iterator_next_finalizing(self_message_iterator,
-                                                   messages, capacity, count);
-  case BTX_FILTER_STATE_FINISHED:
-    return BT_MESSAGE_ITERATOR_CLASS_NEXT_METHOD_STATUS_END;
+    return filter_message_iterator_next_ending(self_message_iterator, messages,
+                                               capacity, count);
```
Comment (Kerilk) [inline template/filter.c.erb:248]:
```suggestion
    return filter_message_iterator_next_ending(self_message_iterator, messages,
                                               capacity, count);
  case BTX_FILTER_STATE_FINISHED:
    eturn BT_MESSAGE_ITERATOR_CLASS_NEXT_METHOD_STATUS_END;
```
his one disappeared in the aftermath.

File: template/filter.c.erb (hunk 5 above)
Comment (TApplencourt) [reply template/filter.c.erb:118]: Correct

**Comment by TApplencourt** [comment]
Will use #72

**Pull request closed without merging**